plugins/database/redshift/redshift.go

```diff
@@ -0,0 +1,519 @@
+func New(lowercaseUsername bool) func() (interface{}, error) {
+	return func() (interface{}, error) {
+		db := new(lowercaseUsername)
+		// Wrap the plugin with middleware to sanitize errors
+		dbType := dbplugin.NewDatabaseErrorSanitizerMiddleware(db, db.SecretValues)
+		return dbType, nil
+	}
+}
+
+func new(lowercaseUsername bool) *RedShift {
+	connProducer := &connutil.SQLConnectionProducer{}
+	connProducer.Type = sqlTypeName
+
+	credsProducer := &credsutil.SQLCredentialsProducer{
+		DisplayNameLen:    8,
+		RoleNameLen:       8,
+		UsernameLen:       63,
+		Separator:         "-",
+		LowercaseUsername: lowercaseUsername,
+	}
+
+	db := &RedShift{
+		SQLConnectionProducer: connProducer,
+		CredentialsProducer:   credsProducer,
+	}
+
+	return db
+}
+
+// Run instantiates a RedShift object, and runs the RPC server for the plugin
+func Run(apiTLSConfig *api.TLSConfig) error {
+	dbType, err := New(true)()
+	if err != nil {
+		return err
+	}
+
+	dbplugin.Serve(dbType.(dbplugin.Database), api.VaultPluginTLSProvider(apiTLSConfig))
+
+	return nil
+}
+
+type RedShift struct {
+	*connutil.SQLConnectionProducer
+	credsutil.CredentialsProducer
+}
+
+func (p *RedShift) Type() (string, error) {
+	return sqlTypeName, nil
+}
+
+func (p *RedShift) getConnection(ctx context.Context) (*sql.DB, error) {
+	db, err := p.Connection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return db.(*sql.DB), nil
+}
+
+// SetCredentials uses provided information to set/create a user in the
+// database. Unlike CreateUser, this method requires a username be provided and
+// uses the name given, instead of generating a name. This is used for creating
+// and setting the password of static accounts, as well as rolling back
+// passwords in the database in the event an updated database fails to save in
+// Vault's storage.
+func (p *RedShift) SetCredentials(ctx context.Context, statements dbplugin.Statements, staticUser dbplugin.StaticUserConfig) (username, password string, err error) {
+	if len(statements.Rotation) == 0 {
+		return "", "", errors.New("empty rotation statements")
+	}
+
+	username = staticUser.Username
+	password = staticUser.Password
+	if username == "" || password == "" {
+		return "", "", errors.New("must provide both username and password")
+	}
+
+	// Grab the lock
+	p.Lock()
+	defer p.Unlock()
+
+	// Get the connection
+	db, err := p.getConnection(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Check if the role exists
+	var exists bool
+	err = db.QueryRowContext(ctx, "SELECT exists (SELECT usename FROM pg_user WHERE usename=$1);", username).Scan(&exists)
+	if err != nil && err != sql.ErrNoRows {
+		return "", "", err
+	}
+
+	// Vault requires the database user already exist, and that the credentials
+	// used to execute the rotation statements has sufficient privileges.
+	stmts := statements.Rotation
+
+	// Start a transaction
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return "", "", err
+	}
+	defer func() {
+		_ = tx.Rollback()
```

Comment:
```suggestion
		tx.Rollback()
```
Might as well just quietly drop the error, we do everywhere else. It would be super nice if we could log these, but alas, no logger!

Reply:
Also, I have a question about these Rollback statements. I see Rollback is called this way in the `mysql` plugin. Is the idea basically that if we succeeded doing the transaction, then Rollback just won't work? I'm wondering if maybe we should add a guard that's like "success" that is false by default, and then it gets set to true if the whole transaction succeeded. Then inside the deferred function, it would be like "if !success, rollback" (pseudocode).